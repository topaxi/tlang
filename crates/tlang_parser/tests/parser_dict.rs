mod common;

#[test]
fn test_dictionary_literal() {
    assert_parser_snapshot!(r"let x = { foo: 1, bar: 2 };");
}

#[test]
fn test_dictionary_literal_trailing_comma() {
    assert_parser_snapshot!(r"let x = { foo: 1, bar: 2, };");
}

#[test]
fn test_dictionary_literal_shorthand() {
    assert_parser_snapshot!(r"let x = { foo, bar };");
}

#[test]
fn test_dictionary_literal_shorthand_trailing_comma() {
    assert_parser_snapshot!(r"let x = { foo, bar, };");
}

#[test]
fn test_function_call_with_dictionary_no_parens() {
    // TODO: Alternatively use ruby like syntax for calling functions with dictionaries
    //       and omit the braces instead of the parens.
    //       As this syntax here might be more desirable for creating dictionaries or structs.
    assert_parser_snapshot!(r"foo { foo: 1, bar: 2 };");
}

#[test]
#[ignore] // TODO: Decide if we want to use this instead of the above syntax.
fn test_fn_call_expression_with_dictionary_argument() {
    // Similar to ruby syntax, we might want to allow omitting curly braces
    // for the last argument if it is a dictionary (or struct).
    // Creation of the struct/dictionary could/should be infered from the function signature.
    assert_parser_snapshot!("bar(x: 1, y: 2)");
}
