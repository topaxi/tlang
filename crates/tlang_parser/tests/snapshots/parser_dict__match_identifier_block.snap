---
source: crates/tlang_parser/tests/parser_dict.rs
expression: "parse! (r\"match foo { _ => 1 }\")"
---
Module(
  id: 2,
  statements: [
    Stmt(
      id: 3,
      kind: Expr(Expr(
        id: 8,
        kind: Match(MatchExpression(
          expression: Expr(
            id: 4,
            kind: Path(Path(
              segments: [
                Ident(
                  name: "foo",
                  span: Span(
                    start: LineColumn(
                      line: 0,
                      column: 6,
                    ),
                    end: LineColumn(
                      line: 0,
                      column: 9,
                    ),
                  ),
                ),
              ],
              res: Res(
                node_id: None,
              ),
              span: Span(
                start: LineColumn(
                  line: 0,
                  column: 6,
                ),
                end: LineColumn(
                  line: 0,
                  column: 9,
                ),
              ),
            )),
            leading_comments: [],
            trailing_comments: [],
            span: Span(
              start: LineColumn(
                line: 0,
                column: 6,
              ),
              end: LineColumn(
                line: 0,
                column: 9,
              ),
            ),
          ),
          arms: [
            MatchArm(
              id: 7,
              pattern: Pat(
                id: 5,
                kind: Wildcard,
                leading_comments: [],
                trailing_comments: [],
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 12,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 13,
                  ),
                ),
              ),
              guard: None,
              expression: Expr(
                id: 6,
                kind: Literal(UnsignedInteger(1)),
                leading_comments: [],
                trailing_comments: [],
                span: Span(
                  start: LineColumn(
                    line: 0,
                    column: 17,
                  ),
                  end: LineColumn(
                    line: 0,
                    column: 18,
                  ),
                ),
              ),
            ),
          ],
        )),
        leading_comments: [],
        trailing_comments: [],
        span: Span(
          start: LineColumn(
            line: 0,
            column: 0,
          ),
          end: LineColumn(
            line: 0,
            column: 20,
          ),
        ),
      )),
      span: Span(
        start: LineColumn(
          line: 0,
          column: 0,
        ),
        end: LineColumn(
          line: 0,
          column: 20,
        ),
      ),
      leading_comments: [],
      trailing_comments: [],
    ),
  ],
  span: Span(
    start: LineColumn(
      line: 0,
      column: 0,
    ),
    end: LineColumn(
      line: 0,
      column: 20,
    ),
  ),
)
